doc/rados/operations/pg-repair.rst

```diff
@@ -1,4 +1,65 @@
+repair" command on it. This page contains commands for diagnosing placement
+groups and the command for repairing placement groups that have become
+inconsistent.
+
+.. highlight:: console
+
+Commands for Diagnosing Placement-group Problems
+================================================
+The commands in this section provide various ways of diagnosing broken placement groups.
+
+The following command provides a high-level (low detail) overview of the health of the ceph cluster::
+
+   # ceph health detail
+
+The following command provides more detail on the status of the placement groups::
+
+   # ceph pg dump --format=json-pretty
+
+The following command lists inconsistent placement groups::
+
+   # rados list-inconsistent-pg {pgid} 
+
+The following command lists inconsistent rados objects::
+
+   # rados list-inconsistent-obj {pgid} 
+ 
+The following command lists inconsistent snapsets in the given placement group::
+
+   # rados list-inconsistent-snapset {pgid}
+
+
+Commands for Repairing Placement Groups
+=======================================
+The form of the command to repair a broken placement group is::
+
+   # ceph pg repair {pgid}
+  
+Where "{pgid}" is the id of the affected placement group.
+
+For example::
+
+   # ceph pg repair 1.4
+
+More Information on Placement Group Repair
+==========================================
+Ceph stores and updates the checksums of objects stored in the cluster. When a scrub is performed on a placement group, the OSD attempts to choose an authoritative copy from among its replicas. Among all of the possible cases, only one case is consistent. After a deep scrub, Ceph calculates the checksum of an object read from the disk and compares it to the checksum previously recorded. If the current checksum and the previously recorded checksums do not match, that is an inconsistency. In the case of replica pools, any mismatch between the checksum of any replica of an object and the checksum of the authoritative copy means that there is an inconsistency.
+
+The "pg repair" command attempts to fix inconsistencies of various kinds. If "pg repair" finds an inconsisent placement group, it attempts to overwrite the digest of the inconsistent copy with the digest of the authoritative copy. If "pg repair" finds an inconsistent replica pool, it marks the inconsistent copy as missing. Recovery, in the case of replica pools, is beyond the scope of "pg repair".
+
+"pg repair" will not solve every problem. Ceph does not automatically repair placement groups when inconsistencies are found in them.
+
+The checksum of an object or an omap is not always available. If an object is updated while a scrub is underway, its checksum is marked invalid. Even though an authoritative copy was available when the scrub was underway, that copy's authoritativeness was questionable. Ceph therefore refuses to repair it even if the user runs "pg repair". This is one of the situations in which manual intervention is necessary.
```

Comment:
> If an object is updated while a scrub is underway, its checksum is marked invalid.

this is not quite right. because the checksum is calculated incrementally. so if an object is updated in its middle, that write operation not only changes the object, it also invalidates its checksum. because we don't think it's worthy to read the **whole** object for recalculating the checksum. but if an operation rewrite the whole object or just append data to it, the checksum is updated accordingly.

Reply:
I've updated the doc. I think that I got this into the text. Let me know if I'm not quite right.